text/3417-dropck-eyepatch-v3.md

```diff
@@ -0,0 +1,239 @@
+        - we recurse into arguments for type parameters marked `#[may_dangle(droppable)]`.
+- Regardless of whether `T` implements `Drop`, recurse into all types *owned* by `T`:
+    - references, raw pointers, function pointers, function items and scalars do not own
+      anything. They can be trivially dropped.
+    - tuples and arrays consider their element types to be owned.
+    - all fields (of all variants) of ADTs are considered owned. We consider all variants
+      for enums. The only exception here is `ManuallyDrop<U>` which is not considered to own `U`.
+      `PhantomData<U>` does not have any fields and therefore also does not consider
+      `U` to be owned.
+    - closures and generators own their captured upvars.
+
+Checking drop impls may error for generic parameters which are known to be incorrectly marked:
+- `#[may_dangle(must_not_use)]` parameters which are recursively owned
+- `#[may_dangle(droppable)]` parameters which are required to be live by a recursively owned type
+
+This cannot catch all misuses, as the parameters can be incorrectly used by the `Drop` impl itself.
+We therefore require the impl to be marked as `unsafe`.
+
+## How this differs from the status quo
+
+Right now there is only the `#[may_dangle]` attribute which skips the generic parameter.
+This is equivalent to the behavior of `#[may_dangle(must_not_use)]` and relies on the recursion
+into types owned by `T` to figure out the correct constraints. This is now explicitly annotated
+using `#[may_dangle(droppable)]`.
+
+`PhantomData<U>` currently considers `U` to be owned while not having drop glue itself. This means
+that `(PhantomData<PrintOnDrop<'s>>, String)` requires `'s` to be live while
+`(PhantomData<PrintOnDrop<'s>>, u32)` does not. This is required for get the
+behavior of `#[may_dangle(droppable)]` for parameters otherwise not owned by adding `PhantomData`
+as a field. One can easily forget this, which caused the
+[unsound](https://github.com/rust-lang/rust/issues/76367)
+[issues](https://github.com/rust-lang/rust/issues/99408) mentioned above.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds a small amount of implementation complexity to the compiler while not not being
+fully checked and therefore requiring `unsafe`.
+
+This RFC does not explicitly exclude stabilizing these two attributes, as they are clearer and far less
+dangerous to use when compared with `#[may_dangle]`. Stabilizing these attributes will make it harder to
+stabilize a more general solution like type state.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The status quo of `#[may_dangle]` and "spooky-dropck-at-a-distance" is far from ideal and has already
+resulted in unsound issues. Documenting the current behavior makes is more difficult to change later
+while not officially documenting it is bound to lead to more issues and confusion going forward.
+It is therefore quite important to improve the status quo.
```

Comment:
I don't buy this as an argument against documenting the status quo, given that this is all unstable.

Reply:
The issue is that it is **not**, or well `#[may_dangle]` is unstable but "spooky-dropck-at-a-distance" affects stable code:

We have the stable behavior that `(PhantomData<PrintOnDrop<'s>>, String)` requires `'s` to be live. Similar for ADTs: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f1f903239c9b472cd9982692205694cf)
```rust
use std::marker::PhantomData;
struct MyType<T, U> {
    // _marker: PhantomData<T>, // errors
    _marker: PhantomData<*const T>, // doesn't error
    other: U,
}

impl<T, U> MyType<T, U> {
    fn mk(_: T, other: U) -> Self {
        MyType {
            _marker: PhantomData,
            other,
        }
    }
}

struct PrintOnDrop<'s>(&'s str);
impl<'s> Drop for PrintOnDrop<'s> {
    fn drop(&mut self) {
        println!("dropping `{}`", self.0)
    }
}

fn main() {
    let x;
    {
        let temp = String::from("temporary");
        x = MyType::mk(PrintOnDrop(temp.as_str()), String::from("needs_drop"));
    }
}
```
While it should be fine™ to change this **stable** behavior even after documenting it as nobody should rely on it, it does make me uncomfortable.

Reply:
Ah I see.
Honestly that seems like cheating to me -- the behavior hasn't been documented, but it has existed on stable for years, and it's not a bug -- this is how things were designed to behave, IIUC. Using our lack of documentation as a loophole feels a bit backhanded.

Reply:
> Honestly that seems like cheating to me -- the behavior hasn't been documented, but it has existed on stable for years, and it's not a bug -- this is how things were designed to behave, IIUC. Using our lack of documentation as a loophole feels a bit backhanded.

do you mean "we should keep the current behavior because it's a breaking change not to"? Or more "whether the current behavior is documented shouldn't change our decision on whether we should change it"?

> this is how things were designed to behave, IIUC

wrt to this I pretty clearly feel like this change is a [clarification of underspecified language semantics](https://rustc-dev-guide.rust-lang.org/bug-fix-procedure.html#what-qualifies-as-a-bug-fix) . I don't think we ever intended to have "does not need `Drop` but when part of something that does, add outlives requirements" as the behavior for `PhantomData`.

> "whether the current behavior is documented shouldn't change our decision on whether we should change it"

I personally disagree with this perspective if that's what you intended to say. By documenting something publicly we declare that the current behavior is intended as is and something users may rely on.

I do believe that the current state here is a bug. Dropping fields separately should be equivalent to dropping them as a pair. Once this lands I intend to add a check that `dropck_outlives_requirements.is_empty()` implies `needs_drop`. i.e. types without drop glue are trivial to drop.

Reply:
> do you mean "we should keep the current behavior because it's a breaking change not to"? Or more "whether the current behavior is documented shouldn't change our decision on whether we should change it"?

The latter. I am totally fine with changing the behavior. But I don't like "we failed to document it" as justification.